### ISSUE TITLE
Nested lists are flattened and can't be accessed using index syntax

### ISSUE THREAD

**Issue by beatcracker** [hashicorp/terraform#19033]
### Terraform Version
```
Terraform v0.11.8
```

### Terraform Configuration Files

```hcl
locals {
  list = [
    ["0", "1"],
    ["2", "3"],
  ]
}

variable index {
  default = 1
}

output item {
  value = "${local.list[var.index]}"
}

```

### Expected Behavior

This should return nested list in output: ```["0", "1"]``` or ```["2", "3"]```

Terraform flattens nested lists, so using index doesn't allow one to access nested lists.

### Actual Behavior

```
Apply complete! Resources: 0 added, 0 changed, 0 destroyed.

Outputs:

item = 1
```

**Comment by mildwonkey** [comment]
Hi @beatcracker!
I am sorry you've run across this problem. It is a known limitation of terraforms somewhat limited type system, and I believe this will behave the way you expect when [terraform 0.12 is released](https://www.hashicorp.com/blog/terraform-0-12-rich-value-types) later this year.

I will label this issue as a bug so we can circle back after the 0.12 release and validate.

**Comment by beatcracker** [comment]
No problem, thanks for the explanation. Currently, I use maps as a workaround:

```hcl
locals {
  map= {
   "0" = ["0", "1"],
   "1" = ["2", "3"],
  }
}
```

**Comment by apparentlymart** [comment]
Hi @beatcracker!

The changes that @mildwonkey mentioned are now merged to master and so I tried your configuration exactly as you wrote it in the v0.12.0-alpha2 prerelease build.

```
$ terraform apply

Apply complete! Resources: 0 added, 0 changed, 0 destroyed.

Outputs:

item = [
  "2",
  "3",
]
```

Success! :tada: 

I think the problem you saw was a quirk of the language parser we used in versions 0.11 and prior. This has been significantly reworked for v0.12 and now has robust support for complex types like this.

The fix is already present in master and ready to be included in the forthcoming v0.12.0 release, so I'm going to close this out. Thanks for reporting this!

**Comment by JiriKovar** [comment]
I just found an workaround for this. It took me a whole day and it was by sheer luck when I was just about to give up. 

```
locals {
  list_of_lists_of_connection_strings = "${concat(azurerm_cosmosdb_account.cosmos_db_that_might_not_exist.*.connection_strings, list(list("")))}"
}

output "primary_connection_string" {
  value = "${element(local.list_of_lists_of_connection_strings[0], 0)}"
}
```

To explain it: 
- cosmos_db_that_might_not_exist is a resource we create only in some environments hence the splat operator
- concat with list of lists with 1 string so that there is at least one element in the outer list and nested lest (therefor providing fallback to empty string)
- in the output there is combination of direct array accessing (retrieving first element from outer array) and "element' function for retrieving the result from inner array. This combination is required, otherwise the module loading fails.
- It must be split to the two parts - into local variable & output, otherwise the module loading complains about ```expected ")" but found "["```
